File: app/src/design-patterns/structural/facade/OrderFacade.ts
import { InventoryService } from "./InventoryService";
import { PaymentService } from "./PaymentService";
import { ShippingService } from "./ShippingService";

export class OrderFacade {
    private inventoryService: InventoryService ;
    private shippingService: ShippingService ;
    private paymentService: PaymentService ;

    // Constructor Injection
    constructor(inventoryService: InventoryService, shippingService: ShippingService , paymentService: PaymentService ) {
        this.inventoryService = inventoryService;
        this.shippingService = shippingService;
        this.paymentService = paymentService;
    }

    public placeOrder(item: string, quantity: number, address: string, amount: number): void {
        this.inventoryService.checkStock(item, quantity);
        this.paymentService.processPayment(amount);
        this.shippingService.arrangeShipping(item, address);
        console.log("Order placed successfully for " + item);
    }
}

File: app/src/design-patterns/structural/facade/OrderFacade.java
package structural.facade;

public class OrderFacade {
    private final InventoryService inventoryService;
    private final ShippingService shippingService;
    private final PaymentService paymentService;

    // Constructor Injection
    public OrderFacade(InventoryService inventoryService, ShippingService shippingService, PaymentService paymentService) {
        this.inventoryService = inventoryService;
        this.shippingService = shippingService;
        this.paymentService = paymentService;
    }

    public void placeOrder(String item, int quantity, String address, double amount) {
        inventoryService.checkStock(item, quantity);
        paymentService.processPayment(amount);
        shippingService.arrangeShipping(item, address);
        System.out.println("Order placed successfully for " + item);
    }
}